sympy/solvers/solveset.py

```diff
@@ -2594,26 +2682,48 @@ def linear_eq_to_matrix(equations, *symbols):
             Eq or Matrix.
             '''))
 
-    A, b = [], []
-    for i, f in enumerate(equations):
-        if isinstance(f, Equality):
-            f = f.rewrite(Add, evaluate=False)
-        coeff_list = linear_coeffs(f, *symbols)
-        b.append(-coeff_list.pop())
-        A.append(coeff_list)
-    A, b = map(Matrix, (A, b))
+    # construct the dictionaries
+    try:
+        eq, c = _linear_eq_to_dict(equations, symbols, strict=strict)
+    except PolyNonlinearError as err:
+        raise NonlinearError(str(err)) from err
+    n, m = shape = len(eq), len(symbols)
+    ix = dict(zip(symbols, range(m)))
+    dod = {row: {ix[k]: d[k] for k in d} for row, d in enumerate(eq)}
+    rhs = [-i for i in c]
+    del c
+    if fmt != '':
+        types = set(map(type, eq.values())) | set(map(type, rhs))
+        dom = None
+        if len(types) == 1:
+            if rhs[0].is_Integer:
+                dom = ZZ
+            elif rhs[0].is_Rational:
+                dom = QQ
+        if dom is None:
+            dom = EXRAW
```

Comment:
I would really like to keep this kind of logic centralised within the RepMatrix class.

Reply:
I have removed this code. It seems like this is something that should be handled in the solver itself -- to pick the right representation in which to do the calculation.